### ISSUE TITLE
Nan loss for classification - LSTM while training

### ISSUE THREAD

**Issue by prabhakaran2812** [keras-team/keras#14490]
This is my code
data_train = pd.read_csv(trainingdata_fn)
x_train = np.array(data_train.iloc[:2020,1:465])
y_train = np.array(data_train.iloc[:2020,-1:])
x_test = np.array(data_train.iloc[2020:2520,1:465])
y_test = np.array(data_train.iloc[2020:2520,-1:])
x_train = x_train.reshape(-1,10,464)
x_test = x_test.reshape(-1,10,464)
y_train = y_train.reshape(-1,10,1)
y_test = y_test.reshape(-1,10,1)

model=Sequential()
model.add(LSTM(10,return_sequences=True,input_shape=(10,464)))
model.add(LSTM(10,return_sequences=True))
model.add(LSTM(10))
model.add(Dense(1))
model.compile(loss=tf.keras.losses.BinaryCrossentropy(),optimizer='adam',metrics=['accuracy'])

model.fit(x_train,y_train,validation_data=(x_test,y_test),epochs=50,batch_size=5,verbose=1)

Im getting Nan loss from epoch 1 through 50

Sample output 
Epoch 1/50
41/41 [==============================] - 7s 51ms/step - loss: nan - accuracy: 0.6654 - val_loss: nan - val_accuracy: 0.5300
Epoch 2/50
41/41 [==============================] - 1s 18ms/step - loss: nan - accuracy: 0.6780 - val_loss: nan - val_accuracy: 0.5300
Epoch 3/50
41/41 [==============================] - 1s 18ms/step - loss: nan - accuracy: 0.6616 - val_loss: nan - val_accuracy: 0.5300
Epoch 4/50

**Comment by ariG23498** [comment]
Hey @prabhakaran2812 
Due to the fact that I did not have the data you trained your model on I tried with **random numbers** but tried keeping the shape similar and targets were binary (0 or 1). I did not find any problem with the working.

## Code
```python
data_train = np.zeros(shape=(2520,466))
data_train[:, :465] = np.random.normal(size=(2520,465))
data_train[:, 465] = np.random.randint(low=0, high=2, size=(2520), dtype=int)

x_train = data_train[:2020,:464]
y_train = data_train[:2020,465]

x_test = data_train[2020:,:464]
y_test = data_train[2020:,465]

x_train = x_train.reshape(-1,10,464)
x_test = x_test.reshape(-1,10,464)
y_train = y_train.reshape(-1,10,1)
y_test = y_test.reshape(-1,10,1)

model=tf.keras.Sequential()
model.add(tf.keras.layers.LSTM(10,return_sequences=True,input_shape=(10,464)))
model.add(tf.keras.layers.LSTM(10,return_sequences=True))
model.add(tf.keras.layers.LSTM(10))
model.add(tf.keras.layers.Dense(1))
model.compile(loss=tf.keras.losses.BinaryCrossentropy(),optimizer='adam',metrics=['accuracy'])

model.fit(x_train,y_train,validation_data=(x_test,y_test),epochs=10,batch_size=5,verbose=1)
```

It would be a good first thing to start getting into the data and scaling them. I think the model might explode with the gradients. Another probable thing that you could do is to check the gradients of the architecture with tensorboard or with any other ML tracking tool. Hope this helps 😄

**Comment by prabhakaran2812** [comment]
@ariG23498 - Thanks for the reply and your effort.
My data is already normalized (0 to 1 - all features). I tried decreasing learning rate but didn't help. Exploring the tensorboard debugging. Its not writing histogram in tensorboard may be that first epoch is throwing Nan. This  is the first time building LSTM. Im missing something fundamental. Any ideas / pointers will be helpful. I will keep exploring tensorboard.

Thanks Again Aritra!

**Comment by prabhakaran2812** [comment]
When I reduced the feature count from 464 to 20, it started giving loss value. Don't understand this behavior.
Any idea @ariG23498
?

**Comment by ariG23498** [comment]
Could you send the input data over?
That would be really good to have to find the real problem here!

**Comment by prabhakaran2812** [comment]
Here you go @ariG23498 . rename the file as from train.txt to train.csv
columns 1-464 are features, last column(binary) is label when i use the full 464 features, loss always nan but when i restrict to first 20 columns its giving some loss and predicting poorly. Let me know what you find. Thanks in advance.

[train.txt](https://github.com/keras-team/keras/files/6073894/train.txt)

**Comment by ariG23498** [comment]
Hey @prabhakaran2812 
Doing a quick EDA on the data
`data_train.isnull().values.any()` I found that the data itself has `NaN`. You need to clean the data first I reckon.
I think you should close this ticket, as this is not a `keras` error.

**Comment by Dragon-GCS** [comment]
Is there some NaN value in your origin input data?

**Comment by prabhakaran2812** [comment]
Yes, but I cleared and tested again, with 400+ features its still giving Nan but i reduce to below 21 it is giving a loss value but the prediction is 0% accurate. My only question is why it is able to give a loss for up to 20 feature but NaN if I increase to 21 or above. I will anyway close the ticket in a day or two.
Thanks @ariG23498 and @Dragon-GCS
